nway/diagnostics.py

```diff
@@ -361,6 +373,201 @@ def plot_all(nway_output_path, fname=None):
         p.close()
 
 
+def create_nway_input_maps(nway_input: dict) -> Tuple[dict, dict]:
+    
+    ophys_expts = nway_input['experiment_containers']['ophys_experiments']
+
+    expt_id_stim_name_map = dict()
+    expt_id_avg_image_map = dict()
+    for expt in ophys_expts:
+        expt_id_stim_name_map[expt['id']] = expt['stimulus_name']
+
+        avg_image_key = 'ophys_average_intensity_projection_image'
+        with PIL.Image.open(expt[avg_image_key]) as im:
+            expt_avg_image = np.array(im)
+        expt_id_avg_image_map[expt['id']] = expt_avg_image
+    
+    expt_id_stim_name_map = {k: v for k, v
+                             in sorted(expt_id_stim_name_map.items(),
+                                       key=lambda x: x[1])}
+    
+    return (expt_id_stim_name_map, expt_id_avg_image_map)
+
+
+def create_nway_summary_df(nway_input: dict,
+                           nway_output: dict) -> pd.DataFrame:
+
+    input_maps = create_nway_input_maps(nway_input)
+    expt_id_stim_name_map, expt_id_avg_image_map = input_maps
+    pairwise_results = nway_output['pairwise_results']
+
+    warped_avg_image_maps = dict()
+    df_list = []
+    for pair in pairwise_results:
+        # Assemble match statistics
+        fixed_expt = pair['fixed_experiment']
+        fixed_expt_stim_name = expt_id_stim_name_map[fixed_expt]
+        
+        moving_expt = pair['moving_experiment']
+        moving_expt_stim_name = expt_id_stim_name_map[moving_expt]
+
+        n_unmatched_fixed = len(pair['unmatched']['fixed'])
+        n_unmatched_moving = len(pair['unmatched']['moving'])
+        n_matches = len(pair['matches'])
+        n_total = n_matches + n_unmatched_fixed + n_unmatched_moving
+        fraction_matched = n_matches / float(n_total)
+        df_list.append([fixed_expt, fixed_expt_stim_name, moving_expt,
+                        moving_expt_stim_name, n_unmatched_fixed,
+                        n_unmatched_moving, n_matches, n_total,
+                        fraction_matched])
+
+        # Recreate warped images
+        transform_matrix = np.array(pair['transform']['matrix'])
+        transform_type = pair['transform']['transform_type']
+        fixed_img = expt_id_avg_image_map[fixed_expt]
+        moving_img = expt_id_avg_image_map[moving_expt]
+
+        moving_warped = imutils.warp_image(moving_img, transform_matrix,
+                                           transform_type, fixed_img.shape)
+        warped_img_key_1 = f"{moving_expt}_to_{fixed_expt}"
+        warped_avg_image_maps[warped_img_key_1] = moving_warped
+        
+        fixed_warped = imutils.warp_image(fixed_img, transform_matrix.T,
```

Comment:
@djkapner 

Also was not 100% certain that transposing the transform_matrix gives you the transform from fixed -> moving.

This is needed to be able to properly iterate over experiments for the 'red green' plots.

Reply:
I see what you're doing here... lower triangular is warping fixed, for example, and upper triangular is warping moving.
I would not have gone that far!
This seems like the right way to do it.
But, I would expect symmetry in these 2 red-green plots and the ssim metric, if this were working correctly.
![image](https://user-images.githubusercontent.com/32312979/101107618-813c4580-3587-11eb-881e-ba595bd2e220.png)

Reply:
oh wait. No, not transpose. inverse. That's probably the problem.

Reply:
Ah, good catch. It's not always the case that a nice 'triangle' of fixed vs moving is formed. See:

![image](https://user-images.githubusercontent.com/43766899/101107928-44248300-3588-11eb-954f-70ac393353e6.png)

Reply:
```
In [36]: T
Out[36]: 
array([[  1. ,   0.1,  23. ],
       [ -0.1,   1. , -43. ],
       [  0. ,   0. ,   1. ]])

In [37]: Tinv = np.linalg.inv(T)

In [38]: T.dot(Tinv)
Out[38]: 
array([[ 1.00000000e+00, -5.49615359e-20,  0.00000000e+00],
       [ 0.00000000e+00,  1.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  1.00000000e+00]])

In [39]: Tinv.dot(T)
Out[39]: 
array([[ 1.00000000e+00,  0.00000000e+00, -3.55271368e-15],
       [ 5.49615359e-20,  1.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  1.00000000e+00]])
```

Reply:
oh, i see what you are saying.
I think the reason for this is that we are iterating through the experiments as they are listed in the input json with `itertools.combinations` and there is some order where this would be nicely triangular, but, not necessarily the order in which you want to plot them.

Reply:
This is now done, switched to using `np.linalg.inv`. Though the SSIM values are still not exactly the same, they are now pretty close.